### PR TITLE
PT-3444: DefaultCollaborator objects with null user cannot be added to any hashed collection [fixed]

### PR DESCRIPTION
**Pull request by a-gvo**

### CODE REVIEW

File: components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java [hunk 1]
```diff
@@ -167,6 +167,32 @@ public void equalsTest() throws ComponentLookupException
...
+    @Test
+    public void equalsWithNullUserTest() throws ComponentLookupException
+    {
+        Collaborator c = new DefaultCollaborator(null, this.access, this.helper);
+        // Equals itself
+        Assert.assertTrue(c.equals(c));
+        // Doesn't equal null
+        Assert.assertFalse(c.equals(null));
+        // Equals an identical collaborator
+        AccessLevel otherAccess = mock(AccessLevel.class);
```
Comment (sdumitriu) [inline components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java]: This isn't used until later, move this declaration on line 184.

File: components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java [hunk 2]
```diff
@@ -167,6 +167,32 @@ public void equalsTest() throws ComponentLookupException
         Assert.assertFalse(c.equals("other"));
     }
 
+    /** Basic tests for {@link Collaborator#equals(Object)}. */
```
Comment (sdumitriu) [inline components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java]: Javadoc is not up to date. Either remove or make it more relevant.

File: components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java [hunk 3]
```diff
@@ -167,6 +167,32 @@ public void equalsTest() throws ComponentLookupException
...
+        Assert.assertFalse(c.equals(null));
+        // Equals an identical collaborator
+        AccessLevel otherAccess = mock(AccessLevel.class);
+        Collaborator other = new DefaultCollaborator(null, this.access, this.helper);
+        Assert.assertTrue(c.equals(other));
+        // Doesn't equal a collaborator with same user but different access
+        other = new DefaultCollaborator(null, otherAccess, this.helper);
+        Assert.assertFalse(c.equals(other));
+        // Doesn't equal a collaborator with same access but different user
+        other = new DefaultCollaborator(new DocumentReference("xwiki", "XWiki", "padams"), this.access, this.helper);
```
Comment (sdumitriu) [inline components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java]: I'd like to see a test in the other direction as well, either `other equals c` here, or add `c equals new collaborator(null, access, helper)` above in `equalsTest`.

Comment (a-gvo) [reply components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java]: Fixed.

File: components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java (hunk 1 above)
Comment (a-gvo) [reply components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java]: Fixed.

File: components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java (hunk 2 above)
Comment (a-gvo) [reply components/patient-access-rules/api/src/test/java/org/phenotips/data/permissions/internal/DefaultCollaboratorTest.java]: Fixed.

**Pull request merged**